cmd/mt-kafka-mdm-report-out-of-order/inputooofinder.go

```diff
@@ -0,0 +1,159 @@
+	cassandraIndex := cassandra.New(cassandra.CliConfig)
+	err := cassandraIndex.InitBare()
+	if err != nil {
+		log.Fatalf("error initializing cassandra index: %s", err.Error())
+		os.Exit(1)
+	}
+
+	metricDefinitions := make([]schema.MetricDefinition, 0)
+	for partition := partitionFrom; (partitionTo == -1 && partition == partitionFrom) || (partitionTo > 0 && partition < partitionTo); partition++ {
+		metricDefinitions = cassandraIndex.LoadPartitions([]int32{int32(partition)}, metricDefinitions, time.Now())
+	}
+
+	tracker := Tracker{}
+	for _, metricDefinition := range metricDefinitions {
+		tracker[metricDefinition.Id] = Track{
+			Name:          metricDefinition.Name,
+			Tags:          metricDefinition.Tags,
+			reorderBuffer: mdata.NewReorderBuffer(reorderWindow, uint32(metricDefinition.Interval), false),
+		}
+	}
+
+	return &inputOOOFinder{
+		prefix: prefix,
+		substr: substr,
+
+		reorderWindow: reorderWindow,
+		tracker:       tracker,
+
+		groupByName:             groupByName,
+		outOfOrderGroupedByName: outOfOrderGroupedByName,
+		duplicatesGroupedByName: duplicatesGroupedByName,
+		groupByTag:              groupByTag,
+		outOfOrderGroupedByTag:  outOfOrderGroupedByTag,
+		duplicatesGroupedByTag:  duplicatesGroupedByTag,
+
+		lock: sync.Mutex{},
+	}
+}
+
+func (ip *inputOOOFinder) incrementGroupings(groupedByName *map[string]int, groupedByTag *map[string]int, track Track) {
+	if ip.groupByName == true {
+		(*groupedByName)[track.Name]++
+	}
+
+	if ip.groupByTag != "" {
+		for _, tag := range track.Tags {
+			kv := strings.Split(tag, "=")
+			if len(kv) != 2 {
+				log.Errorf("unexpected tag encoding tag=%q", tag)
+				continue
+			}
+			if kv[0] == ip.groupByTag {
+				(*groupedByTag)[kv[1]]++
+			}
+		}
+	}
+}
+
+func (ip *inputOOOFinder) processTrack(metricKey schema.MKey, metricTime int64, track Track, partition int32) {
+	if ip.prefix != "" && !strings.HasPrefix(track.Name, ip.prefix) {
+		return
+	}
+	if ip.substr != "" && !strings.Contains(track.Name, ip.substr) {
+		return
+	}
+
+	_, err := track.reorderBuffer.Add(uint32(metricTime), 0) // ignore value
+	if err == errors.ErrMetricTooOld {
+		ip.incrementGroupings(ip.outOfOrderGroupedByName, ip.outOfOrderGroupedByTag, track)
+	} else if err == errors.ErrMetricNewValueForTimestamp {
+		ip.incrementGroupings(ip.duplicatesGroupedByName, ip.duplicatesGroupedByTag, track)
+	} else if err != nil {
+		log.Errorf("failed to add metric with name=%q and timestamp=%d from partition=%d to reorder buffer: %s", track.Name, metricTime, partition, err)
+	}
+}
+
+func (ip *inputOOOFinder) ProcessMetricData(metric *schema.MetricData, partition int32) {
+	metricKey, err := schema.MKeyFromString(metric.Id)
+	if err != nil {
+		log.Errorf("failed to get metric key from id=%v: %s", metric.Id, err.Error())
+		return
+	}
+
+	ip.lock.Lock()
+	defer ip.lock.Unlock()
```

Comment:
I worry that this locking might be too coarse and will reduce the series we can analyze. Probably not a big deal in the first iteration though.